src/providers/aws.rs

```diff
@@ -0,0 +1,887 @@
+//! AWS backend for tsunami.
+//!
+//! The primary `impl Launcher` type is [`Launcher`].
+//! It internally uses the lower-level, region-specific [`RegionLauncher`].
+//! Both these types use [`Setup`] as their descriptor type.
+//!
+//! This implementation uses [defined duration](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/spot-requests.html#fixed-duration-spot-instances)
+//! instances.
+//!
+//! # Example
+//! ```rust,no_run
+//! use tsunami::providers::{aws, Launcher};
+//! use tsunami::TsunamiBuilder;
+//!
+//! let mut b = TsunamiBuilder::default();
+//! b.add("my machine", aws::Setup::default()).unwrap();
+//! let mut l = aws::Launcher::default();
+//! b.spawn(&mut l).unwrap();
+//! let vms = l.connect_all().unwrap();
+//! let my_machine = vms.get("my machine").unwrap();
+//! let (stdout, stderr) = my_machine.ssh.as_ref().unwrap().cmd("echo \"Hello, EC2\"").unwrap();
+//! println!("{}", stdout);
+//! ```
+
+use crate::ssh;
+use crate::Machine;
+use educe::Educe;
+use failure::{Error, ResultExt};
+use itertools::Itertools;
+use rusoto_core::request::HttpClient;
+use rusoto_core::{DefaultCredentialsProvider, ProvideAwsCredentials, Region};
+use rusoto_ec2::Ec2;
+use std::collections::HashMap;
+use std::io::Write;
+use std::sync::Arc;
+use std::{thread, time};
+
+struct UbuntuAmi(String);
+
+impl From<Region> for UbuntuAmi {
```

Comment:
Maybe move some of this business to the bottom of the file like we did for Azure?